File: bindings/python/perf/multi/perf_multi_spot_reqrep_client.py
import os
import sys
import threading
import time
from queue import SimpleQueue

import zlink

from perf_multi_common import (
    TOPIC,
    apply_multi_auto_hwm_msg_unit,
    apply_multi_spot_node_admission,
    benchmark_endpoint,
    bind_spot_node_endpoint,
    benchmark_run_id,
    configure_multi_tls_client,
    configure_multi_tls_server,
    is_active_message,
    latency_ns_from_message,
    metric_payload_data,
    parse_client_args,
    perf_client_context,
    print_result_lines,
    publish_control_payload,
    receive_control_payload,
    new_spot_poller,
    resolve_multi_connect_ready_timeout_ms,
    resolve_multi_spot_control_settle_s,
    resolve_multi_spot_ready_settle_s,
    result_metrics,
    stamp_payload,
    wait_control_readable_until,
    wait_spot_peer_connected,
)


SERVER_NODE_RID = b"SPOT-REQREP-SERVER-NODE"
SERVER_SPOT_RID = b"SPOT-REQREP-SERVER-SPOT"


def _active_slot_limit(total_slots, msg_size):
    if msg_size >= 131072:
        return min(total_slots, 8)
    if msg_size >= 65536:
        return min(total_slots, 32)
    return total_slots


def main(argv=None):
    args = parse_client_args(argv or sys.argv[1:], pattern="spot_reqrep")
    if not args.control_endpoint:
        raise SystemExit("spot reqrep client expects --endpoint and --control-endpoint")

    run_id = benchmark_run_id()
    latencies = []
    ready_settle_s = resolve_multi_spot_ready_settle_s()
    control_settle_s = resolve_multi_spot_control_settle_s()
    handshake_timeout_s = resolve_multi_connect_ready_timeout_ms() / 1000.0
    runner_start = threading.Event()
    control_connected = threading.Event()

    def stdin_loop():
        for line in sys.stdin:
            text = line.strip()
            if not text:
                continue
            if text == f"START,{args.msg_size}":
                runner_start.set()
            elif text.startswith("CONTROL_CONNECTED,"):
                control_connected.set()
            elif text in {"STOP", "QUIT"}:
                return

    threading.Thread(target=stdin_loop, daemon=True).start()

    with perf_client_context() as ctx:
        apply_multi_auto_hwm_msg_unit(ctx, args.msg_size)
        data_node = zlink.create_spot_node(ctx)
        control_node = zlink.create_spot_node(ctx)
        configure_multi_tls_server(data_node, args.transport)
        configure_multi_tls_client(data_node, args.transport)
        configure_multi_tls_server(control_node, args.transport)
        configure_multi_tls_client(control_node, args.transport)
        apply_multi_spot_node_admission(data_node, control_node)
        data_node.set_routing_id(b"SPOT-REQREP-CLIENT-NODE")
        control_node.set_routing_id(b"SPOT-REQREP-CONTROL-CLIENT-NODE")

        control_pub = control_node.create_spot()
        control_sub = control_node.create_spot()
        control_pub.set_routing_id(b"SPOT-REQREP-CONTROL-CLIENT-PUB")
        control_sub.set_routing_id(b"SPOT-REQREP-CONTROL-CLIENT-SUB")
        control_sub.set_subscription(TOPIC)
        control_endpoint = bind_spot_node_endpoint(
            control_node, args.transport, "multi-spot-reqrep-control-client"
        )
        control_node.connect_peer(args.control_endpoint)
        print(f"CLIENT_CONTROL_ENDPOINT,{control_endpoint}", flush=True)

        data_node.set_router_bind(
            benchmark_endpoint(args.transport, "multi-spot-reqrep-client-router")
        )
        data_endpoint_local = bind_spot_node_endpoint(
            data_node, args.transport, "multi-spot-reqrep-client"
        )
        data_node.connect_peer(args.endpoint)
        spots = []
        payloads = [bytearray(args.msg_size) for _ in range(args.clients)]
        seq = 0
        for index in range(args.clients):
            spot = data_node.create_spot()
            spot.set_routing_id(f"spot-req-client-spot-{index}".encode("ascii"))
            spots.append(spot)
        poller = zlink.create_poller()
        poll_events = zlink.create_poll_events(max(1, len(spots)))
        for index, spot in enumerate(spots):
            poller.add_socket(spot, zlink.PollEventFlag.POLLCOMPLETION, index)

        if not control_connected.wait(timeout=handshake_timeout_s):
            raise RuntimeError("runner control-connected handshake timeout")

        time.sleep(ready_settle_s)
        time.sleep(control_settle_s)
        if not publish_control_payload(
            control_pub,
            f"DATA_ENDPOINT,{data_endpoint_local}",
            timeout_s=handshake_timeout_s,
        ):
            raise RuntimeError("spot reqrep data endpoint publish timeout")
        time.sleep(control_settle_s)
        if not wait_spot_peer_connected(data_node, handshake_timeout_s):
            raise RuntimeError("spot reqrep data peer connection timeout")
        publish_control_payload(control_pub, "CONNECTED", timeout_s=handshake_timeout_s)
        time.sleep(control_settle_s)

        if not publish_control_payload(
            control_pub,
            f"READY_COUNT,{args.msg_size},{args.clients}",
            timeout_s=handshake_timeout_s,
        ):
            raise RuntimeError("spot reqrep ready publish timeout")
        print(f"CLIENT_READY,{args.msg_size}", flush=True)

        if not runner_start.wait(timeout=handshake_timeout_s):
            raise RuntimeError("spot reqrep runner start handshake timeout")

        direct_start_deadline = time.perf_counter() + handshake_timeout_s
        control_poller, control_events = new_spot_poller(
            control_sub, zlink.PollEventFlag.POLLIN
        )
        direct_started = False
        while time.perf_counter() < direct_start_deadline:
            payload_text = receive_control_payload(control_sub)
            if payload_text is None:
                wait_control_readable_until(
                    control_poller, control_events, direct_start_deadline
                )
                continue
            if payload_text == f"START,{args.msg_size}":
                direct_started = True
                break
        if not direct_started:
            raise RuntimeError("spot reqrep direct start handshake timeout")
        control_poller.close()

        active_deadline = time.perf_counter() + args.duration
        waiting = [False] * len(spots)
        lock = threading.Lock()
        latency_queue = SimpleQueue()
        active_slots = _active_slot_limit(len(spots), args.msg_size)
        while time.perf_counter() < active_deadline:
            submitted_any = False
            for index in range(active_slots):
                spot = spots[index]
                with lock:
                    if waiting[index]:
                        continue
                    waiting[index] = True
                seq += 1
                payload = stamp_payload(
                    payloads[index],
                    phase=1,
                    run_id=run_id,
                    seq=seq,
                )

                def on_reply(result, messages, slot_index=index):
                    try:
                        if result != zlink.RequestResult.OK or not messages:
                            return
                        data = metric_payload_data(messages[0].data)
                        if (
                            time.perf_counter() < active_deadline
                            and is_active_message(
                                data,
                                expected_msg_size=args.msg_size,
                                run_id=run_id,
                            )
                        ):
                            latency = latency_ns_from_message(data)
                            if latency is not None:
                                latency_queue.put(latency / 2.0)
                    finally:
                        with lock:
                            waiting[slot_index] = False

                submitted = (
                    False
                )
                try:
                    submitted = (
                        spot.request_to_spot(SERVER_NODE_RID, SERVER_SPOT_RID)
                        .message(bytes(payload))
                        .timeout(0.2)
                        .flags(zlink.SendFlags.DONT_WAIT)
                        .submit(on_reply)
                    )
                except zlink.SubmitError as exc:
                    if exc.result not in (
                        zlink.SubmitResult.BACKPRESSURED,
                        zlink.SubmitResult.NOT_CONNECTED,
                    ):
                        raise
                if submitted:
                    submitted_any = True
                else:
                    with lock:
                        waiting[index] = False
            while not latency_queue.empty():
                latencies.append(latency_queue.get())
            if submitted_any:
                continue
            remaining_ms = int((active_deadline - time.perf_counter()) * 1000)
            if remaining_ms <= 0:
                break
            poller.wait(poll_events, min(50, remaining_ms))

        drain_deadline = time.perf_counter() + 1.0
        while any(waiting[:active_slots]) and time.perf_counter() < drain_deadline:
            poller.wait(poll_events, 50)
            while not latency_queue.empty():
                latencies.append(latency_queue.get())

        while not latency_queue.empty():
            latencies.append(latency_queue.get())

        if not latencies:
            raise RuntimeError(
                "multi spot reqrep benchmark did not receive any active reply"
            )
        metrics = result_metrics(
            count=len(latencies),
            msg_size=args.msg_size,
            elapsed_s=max(args.duration, 0.001),
            latencies_ns=latencies,
            bandwidth_multiplier=2.0,
        )
        print_result_lines("MULTI_SPOT_REQREP", args.transport, args.msg_size, metrics)


if __name__ == "__main__":
    main()
    sys.stdout.flush()
    sys.stderr.flush()
    os._exit(0)

File: bindings/rust/perf/multi/src/perf_multi_spot_reqrep_client.rs
#[path = "perf_common.rs"]
mod common;

use std::io::{self, BufRead, Write};
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::{mpsc, Arc};
use std::thread;
use std::time::{Duration, Instant};

use zlink::POLLCOMPLETION;
use zlink::{
    Message, PollEvent, Poller, RecvFlags, RecvResult, RoutingId, SendFlags, Spot, SpotNode,
    SubmitResult, TopicMessage,
};

const PATTERN: &str = "MULTI_SPOT_REQREP";
const TOPIC: &str = "bench";
const SERVER_NODE_RID: &[u8] = b"SPOT-REQREP-SERVER-NODE";
const SERVER_SPOT_RID: &[u8] = b"SPOT-REQREP-SERVER-SPOT";

enum ClientEvent {
    RunnerControlConnected,
    RunnerStart,
    Stop,
}

fn env_u64(name: &str, default: u64) -> u64 {
    std::env::var(name)
        .ok()
        .and_then(|value| value.parse().ok())
        .unwrap_or(default)
}

fn setup_tls_server(node: &SpotNode, transport: &str) {
    if matches!(transport, "tls" | "wss") {
        let tls = common::resolve_perf_tls_paths().expect("TLS certs not found");
        node.set_tls_server(&tls.cert, &tls.key, false)
            .expect("spot tls server");
    }
}

fn setup_tls_client(node: &SpotNode, transport: &str) {
    if matches!(transport, "tls" | "wss") {
        let tls = common::resolve_perf_tls_paths().expect("TLS certs not found");
        node.set_tls_client(&tls.ca, "localhost", false)
            .expect("spot tls client");
    }
}

fn active_spot_reqrep_slot_limit(total_slots: usize, msg_size: usize) -> usize {
    if msg_size >= 131072 {
        return total_slots.min(8);
    }
    if msg_size >= 65536 {
        return total_slots.min(32);
    }
    total_slots
}

fn control_payload(control_sub: &Spot) -> Option<String> {
    let mut message = TopicMessage::empty();
    match control_sub.subscribe(&mut message, RecvFlags::DONT_WAIT) {
        Ok(true) => {
            let data = common::message_payload(message.parts());
            Some(String::from_utf8_lossy(data).into_owned())
        }
        Ok(false) => None,
        Err(err) if err.code() == RecvResult::NoData => None,
        Err(err) => panic!("control subscribe failed: {err}"),
    }
}

fn publish_control(control_pub: &Spot, payload: &str, timeout: Duration) -> bool {
    let deadline = Instant::now() + timeout;
    while Instant::now() < deadline {
        match control_pub
            .publish(TOPIC)
            .message(Message::try_from(payload.as_bytes()).expect("control message"))
            .flags(SendFlags::DONT_WAIT)
            .submit()
        {
            Ok(_) => return true,
            Err(err) if err.code() == SubmitResult::Backpressured => {
                common::poll_idle_until(deadline, Duration::from_millis(10));
            }
            Err(err) => panic!("control publish failed: {err}"),
        }
    }
    false
}

fn main() {
    let args = common::MultiArgs::parse();
    let settings = common::MultiSettings::from_env();
    let (data_endpoint, control_endpoint) = args
        .endpoint
        .split_once(',')
        .expect("spot reqrep client expects data,control endpoint");
    let data_endpoint = data_endpoint.to_string();
    let control_endpoint = control_endpoint.to_string();
    let ready_settle = Duration::from_millis(env_u64("PERF_MULTI_SPOT_READY_SETTLE_MS", 1000));
    let control_settle = Duration::from_millis(env_u64("PERF_MULTI_SPOT_CONTROL_SETTLE_MS", 25));
    let ready_timeout = common::resolve_multi_connect_ready_timeout();
    let (event_tx, event_rx) = mpsc::channel::<ClientEvent>();

    {
        let event_tx = event_tx.clone();
        let msg_size = args.msg_size;
        thread::spawn(move || {
            let stdin = io::stdin();
            for line in stdin.lock().lines() {
                let line = line.unwrap_or_default();
                let text = line.trim();
                if text == format!("START,{msg_size}") {
                    let _ = event_tx.send(ClientEvent::RunnerStart);
                } else if text.starts_with("CONTROL_CONNECTED,") {
                    let _ = event_tx.send(ClientEvent::RunnerControlConnected);
                } else if matches!(text, "STOP" | "QUIT") {
                    let _ = event_tx.send(ClientEvent::Stop);
                    return;
                }
            }
        });
    }

    let ctx = common::perf_client_context();
    common::apply_multi_auto_hwm_msg_unit(&ctx, args.msg_size);
    let data_node = SpotNode::new(&ctx).expect("spot data node");
    let control_node = SpotNode::new(&ctx).expect("spot control node");
    setup_tls_server(&data_node, &args.transport);
    setup_tls_client(&data_node, &args.transport);
    setup_tls_server(&control_node, &args.transport);
    setup_tls_client(&control_node, &args.transport);
    common::apply_multi_spot_node_admission(&data_node, &settings);
    common::apply_multi_spot_node_admission(&control_node, &settings);
    data_node
        .set_routing_id(&RoutingId::from(b"SPOT-REQREP-CLIENT-NODE"))
        .expect("data rid");

    let control_pub = control_node.create_spot().expect("control pub");
    let control_sub = control_node.create_spot().expect("control sub");
    control_sub.set_subscription(TOPIC).expect("control sub");
    let Some(control_bind) =
        common::benchmark_endpoint(PATTERN, &args.transport, "multi-spot-reqrep-control-client")
    else {
        return;
    };
    control_node
        .set_pub_bind(&control_bind)
        .expect("control bind");
    control_node
        .connect_peer(&control_endpoint)
        .expect("connect server control");
    let client_control_endpoint = control_node.last_endpoint().unwrap_or(control_bind);
    println!("CLIENT_CONTROL_ENDPOINT,{client_control_endpoint}");
    io::stdout().flush().ok();

    let Some(data_bind) =
        common::benchmark_endpoint(PATTERN, &args.transport, "multi-spot-reqrep-client")
    else {
        return;
    };
    let Some(data_router_bind) =
        common::benchmark_endpoint(PATTERN, &args.transport, "multi-spot-reqrep-router-client")
    else {
        return;
    };
    data_node
        .set_router_bind(&data_router_bind)
        .expect("client data router bind endpoint");
    data_node
        .set_pub_bind(&data_bind)
        .expect("client data bind");
    let data_endpoint_local = data_node.last_endpoint().unwrap_or(data_bind);
    data_node
        .connect_peer(&data_endpoint)
        .expect("connect data");

    let server_node_rid = RoutingId::from(SERVER_NODE_RID);
    let server_spot_rid = RoutingId::from(SERVER_SPOT_RID);
    let mut latency = common::LatencyStats::new();
    let mut spots: Vec<Box<Spot>> = Vec::with_capacity(settings.clients);
    let mut payloads = Vec::with_capacity(settings.clients);
    let mut seqs = vec![1u64; settings.clients];
    let mut waiting = Vec::with_capacity(settings.clients);
    for index in 0..settings.clients {
        let spot = Box::new(data_node.create_spot().expect("spot"));
        spot.set_routing_id(&RoutingId::from(
            format!("SPOT-REQREP-CLIENT-SPOT-{index}").as_bytes(),
        ))
        .expect("spot rid");
        payloads.push(vec![0u8; args.msg_size.max(common::HEADER_SIZE)]);
        waiting.push(Arc::new(AtomicBool::new(false)));
        spots.push(spot);
    }
    let poller = Poller::new().expect("spot reqrep completion poller");
    let mut poll_events = vec![PollEvent::default(); settings.clients.max(1)];
    for (index, spot) in spots.iter().enumerate() {
        poller
            .add_socket(spot.as_ref(), POLLCOMPLETION, index)
            .expect("spot reqrep completion poller add");
    }

    if !matches!(
        event_rx.recv_timeout(ready_timeout),
        Ok(ClientEvent::RunnerControlConnected)
    ) {
        panic!("spot reqrep client control connection timeout");
    }
    thread::sleep(ready_settle);
    thread::sleep(control_settle);
    if !publish_control(
        &control_pub,
        &format!("DATA_ENDPOINT,{data_endpoint_local}"),
        ready_timeout,
    ) {
        panic!("spot reqrep data endpoint publish timeout");
    }
    thread::sleep(control_settle);
    if !common::wait_spot_peer_connected(&data_node, ready_timeout) {
        panic!("spot reqrep data peer connection timeout");
    }
    let _ = publish_control(&control_pub, "CONNECTED", ready_timeout);
    thread::sleep(control_settle);

    if !publish_control(
        &control_pub,
        &format!("READY_COUNT,{},{}", args.msg_size, settings.clients),
        ready_timeout,
    ) {
        panic!("spot reqrep ready publish timeout");
    }
    println!("CLIENT_READY,{}", args.msg_size);
    io::stdout().flush().ok();

    if !matches!(
        event_rx.recv_timeout(ready_timeout),
        Ok(ClientEvent::RunnerStart)
    ) {
        panic!("spot reqrep runner start handshake timeout");
    }
    let control_read_poller = common::control_read_poller(&control_sub);
    let mut control_events = vec![PollEvent::default(); 1];
    let direct_deadline = Instant::now() + ready_timeout;
    let mut direct_started = false;
    while Instant::now() < direct_deadline {
        if let Some(payload) = control_payload(&control_sub) {
            if payload == format!("START,{}", args.msg_size) {
                direct_started = true;
                break;
            }
        }
        common::wait_control_readable_until(
            &control_read_poller,
            &mut control_events,
            direct_deadline,
        );
    }
    if !direct_started {
        panic!("spot reqrep direct start handshake timeout");
    }

    let deadline = Instant::now() + Duration::from_secs(settings.duration_seconds);
    let (latency_tx, latency_rx) = mpsc::channel::<f64>();
    let active_slots = active_spot_reqrep_slot_limit(spots.len(), args.msg_size);
    let msg_size = args.msg_size;
    while Instant::now() < deadline {
        let mut submitted = false;
        for index in 0..active_slots {
            if waiting[index].load(Ordering::Acquire) {
                continue;
            }
            common::encode_header(
                &mut payloads[index],
                common::PHASE_ACTIVE,
                args.msg_size as u32,
                seqs[index],
            );
            let request = Message::try_from(&payloads[index]).expect("request");
            let waiting_flag = Arc::clone(&waiting[index]);
            let tx = latency_tx.clone();
            waiting_flag.store(true, Ordering::Release);
            let submit = spots[index]
                .request_to_spot(server_node_rid.clone(), server_spot_rid.clone())
                .message(request)
                .timeout(Duration::from_millis(
                    settings.receive_timeout_ms.max(settings.send_timeout_ms),
                ))
                .flags(SendFlags::DONT_WAIT)
                .submit(move |result| {
                    waiting_flag.store(false, Ordering::Release);
                    let Ok(reply) = result else {
                        return;
                    };
                    let data = common::message_payload(&reply);
                    if Instant::now() <= deadline && common::is_valid_active_message(data, msg_size)
                    {
                        let sent_ts_ns = common::decode_sent_ts_ns(data);
                        let now_ns = common::now_ns();
                        if sent_ts_ns > 0 && now_ns >= sent_ts_ns as u64 {
                            let _ = tx.send((now_ns - sent_ts_ns as u64) as f64 / 2.0);
                        }
                    }
                });
            match submit {
                Ok(()) => {
                    seqs[index] += 1;
                    submitted = true;
                }
                Err(_) => {
                    waiting[index].store(false, Ordering::Release);
                }
            }
        }
        while let Ok(value) = latency_rx.try_recv() {
            latency.record_ns(value);
        }
        if !submitted {
            let remaining = deadline.saturating_duration_since(Instant::now());
            let wait_ms = remaining.min(Duration::from_millis(50)).as_millis() as i64;
            if wait_ms <= 0 {
                break;
            }
            let _ = poller.wait(&mut poll_events, wait_ms);
        }
    }
    while let Ok(value) = latency_rx.try_recv() {
        latency.record_ns(value);
    }
    let pending_drain_deadline = Instant::now()
        + Duration::from_millis(
            1000.max(settings.receive_timeout_ms.max(settings.send_timeout_ms) * 4),
        );
    while waiting
        .iter()
        .take(active_slots)
        .any(|slot| slot.load(Ordering::Acquire))
        && Instant::now() < pending_drain_deadline
    {
        let remaining = pending_drain_deadline.saturating_duration_since(Instant::now());
        let wait_ms = remaining.min(Duration::from_millis(50)).as_millis() as i64;
        if wait_ms <= 0 {
            break;
        }
        let _ = poller.wait(&mut poll_events, wait_ms);
        while let Ok(value) = latency_rx.try_recv() {
            latency.record_ns(value);
        }
    }

    common::print_result(
        PATTERN,
        &args.transport,
        args.msg_size,
        settings.duration_seconds,
        &latency.finish(),
    );
}
